### PR TITLE
147132 - BS Request and Download Exchange set

### PR DESCRIPTION
**Pull request by shailen14599**
This PR consists changes of [PBI -147132](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/147132)

- Code changes for request, download and extraction of exchange set
- Unit test cases
- Additional temporary code changes for zip and upload the exchange set
- Testing will be covered under [Product Backlog Item 150897](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/150897): Testing : BESS BS - Request, wait/poll and download exchange set

### CODE REVIEW

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs [hunk 1]
```diff
@@ -112,27 +114,42 @@ private static HostBuilder BuildHostConfiguration()
...
+                 serviceCollection.AddScoped<IFssService, FssService>();
+                 serviceCollection.AddHttpClient();
+                 serviceCollection.AddTransient<IEssApiClient, EssApiClient>();
+                 serviceCollection.AddTransient<IFssApiClient, FssApiClient>();
+
+                 serviceCollection.AddScoped<IFileSystemHelper, FileSystemHelper>();
+                 serviceCollection.AddScoped<IFileSystem, FileSystem>();
+                 serviceCollection.AddScoped<IZipHelper, ZipHelper>();
+                 serviceCollection.AddScoped<IFileUtility, FileUtility>();
+
```
Comment (rsj-github) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs:146]: wouldn't this be neater to remove the clutter and abstract

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 2]
```diff
@@ -1,22 +1,222 @@
...
+            if (configQueueMessage.Type == BessType.BASE.ToString())
+                exchangeSetResponseModel = await essService.PostProductIdentifiersData((List<string>)configQueueMessage.EncCellNames, configQueueMessage.ExchangeSetStandard);
+            else if (configQueueMessage.Type == BessType.UPDATE.ToString() ||
+                     configQueueMessage.Type == BessType.CHANGE.ToString())
+            {
+                ProductVersionsRequest productVersionsRequest = GetProductVersionDetails(configQueueMessage.EncCellNames);
+                exchangeSetResponseModel = await essService.GetProductDataProductVersions(productVersionsRequest, configQueueMessage.ExchangeSetStandard);
+            }
+            return CommonHelper.ExtractBatchId(exchangeSetResponseModel.Links.ExchangeSetBatchDetailsUri.Href);
+        }
```
Comment (rsj-github) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:78]: this this do any validation of what the initial list of cells which were requested to the cells fetched from the essService

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/FssService.cs [hunk 3]
```diff
@@ -56,6 +56,11 @@ public async Task<FssBatchStatus> CheckIfBatchCommitted(string batchId, RequestT
                 batchStatusPollingCutoffTime = double.Parse(_fssApiConfiguration.Value.BatchStatusPollingCutoffTime);
                 batchStatusPollingDelayTime = int.Parse(_fssApiConfiguration.Value.BatchStatusPollingDelayTime);
             }
+            else if (requestType.Equals(RequestType.BESS))
+            {
+                batchStatusPollingCutoffTime = double.Parse(_fssApiConfiguration.Value.BatchStatusPollingCutoffTimeForBES);
+                batchStatusPollingDelayTime = int.Parse(_fssApiConfiguration.Value.BatchStatusPollingDelayTimeForBES);
+            }
```
Comment (rsj-github) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/FssService.cs:63]: this is not in the stryker report or at least i don't see the stryker report for the latest execution

**Review by rsj-github — CHANGES_REQUESTED** [review]
left some comments but noticed in general that the stryker report seems to indicated we are missing the targets. Of course i maybe wrong and would be pleased if this can be explained.

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/UKHO.BESS.BuilderService.csproj [hunk 4]
```diff
@@ -6,10 +6,10 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
-    <NoWarn>1701;1702;8600;8602;8618</NoWarn>
+    <NoWarn>1701;1702;8600;8602;8618;8604</NoWarn>
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/UKHO.BESS.BuilderService.csproj:9]: What error is being suppressed here?

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs (hunk 1 above)
Comment (DanOwenUKHO) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Program.cs:146]: Hi Ray, this is a standard service collection setup format in program.cs, no change required here

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 5]
```diff
@@ -1,22 +1,222 @@
...
+            #endregion TemporaryUploadCode
 
             return "Exchange Set Created Successfully";
         }
+
+        private async Task<string> RequestExchangeSetAsync(ConfigQueueMessage configQueueMessage)
+        {
+            ExchangeSetResponseModel exchangeSetResponseModel = new();
+            if (configQueueMessage.Type == BessType.BASE.ToString())
+                exchangeSetResponseModel = await essService.PostProductIdentifiersData((List<string>)configQueueMessage.EncCellNames, configQueueMessage.ExchangeSetStandard);
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:70]: Please add scoping braces '{' for the if + else if scopes

**Review by DanOwenUKHO — COMMENTED** [review]
Couple of small comments

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs (hunk 5 above)
Comment (Madhav15303) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:71]: resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/UKHO.BESS.BuilderService.csproj (hunk 4 above)
Comment (shirinbt14926) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/UKHO.BESS.BuilderService.csproj:9]: 8604 warning is suppressed for null reference argument  for parsing the correlationId

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 6]
```diff
@@ -1,22 +1,224 @@
...
+        {
+            GetBatchResponseModel batchDetail = await fssService.GetBatchDetails(essBatchId);
+            List<FssBatchFile> batchFiles = batchDetail.Files.Select(a => new FssBatchFile { FileName = a.Filename, FileLink = a.Links.Get.Href, FileSize = a.FileSize }).ToList();
+
+            if (batchFiles.Any() && !batchFiles.Any(f => f.FileName.ToLower().Contains("error")))
+            {
+                return batchFiles;
+            }
+
+            logger.LogError(EventIds.ErrorFileFoundInBatch.ToEventId(), "Either no files found or error file found in batch with BatchID - {BatchID} | {DateTime} | _X-Correlation-ID:{CorrelationId}", essBatchId, DateTime.UtcNow, CommonHelper.CorrelationID);
```
Comment (shirinbt14926) [inline UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs:128]:
```suggestion
            logger.LogError(EventIds.ErrorFileFoundInBatch.ToEventId(), "Either no files found in batch or error file found in batch with BatchID - {BatchID} | {DateTime} | _X-Correlation-ID:{CorrelationId}", essBatchId, DateTime.UtcNow, CommonHelper.CorrelationID);
```

**Review by shirinbt14926 — COMMENTED** [review]
Few comments to have a look

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs (hunk 6 above)
Comment (Madhav15303) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 7]
```diff
@@ -1,22 +1,224 @@
...
+            {
+                string batchId = await fssService.CreateBatch(batchType);
+                IEnumerable<string> filePaths = fileSystemHelper.GetFiles(downloadPath, fileExtension, SearchOption.TopDirectoryOnly);
+                UploadBatchFiles(filePaths, batchId, batchType);
+                isCommitted = await fssService.CommitBatch(batchId, filePaths, batchType);
+                logger.LogInformation(EventIds.CreateBatchCompleted.ToEventId(), "Batch added to FSS. BatchId: {batchId} and status: {isCommitted}", batchId, isCommitted);
+            }
+            catch (Exception ex)
+            {
+                logger.LogError(EventIds.CreateBatchFailed.ToEventId(), "Batch create failed with Exception : {ex}", ex);
```
[the remark this answers is not in the document]
Comment (Madhav15303) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: resolved

File: UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs [hunk 8]
```diff
@@ -1,22 +1,224 @@
...
+            });
+        }
+
+        private async Task<bool> CreateBessBatchAsync(string downloadPath, string fileExtension, Batch batchType)
+        {
+            bool isCommitted;
+            try
+            {
+                string batchId = await fssService.CreateBatch(batchType);
+                IEnumerable<string> filePaths = fileSystemHelper.GetFiles(downloadPath, fileExtension, SearchOption.TopDirectoryOnly);
```
[the remark this answers is not in the document]
Comment (Madhav15303) [reply UKHO.PeriodicOutputService/UKHO.BESS.BuilderService/Services/BuilderService.cs]: resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/FssService.cs (hunk 3 above)
Comment (Madhav15303) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Common/Services/FssService.cs:64]: added required changes.

**Review by shirinbt14926 — APPROVED** [review]
LGTM. Thanks

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by DanOwenUKHO — APPROVED** [review]

**Review by harshal11869 — APPROVED** [review]
LGTM.

**Review by rsj-github — APPROVED** [review]

**Pull request merged**